[api] improve config options

ashulin opened the pull request:
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist

  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/incubator-seatunnel/issues).

  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.

  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.

-->

## Purpose of this pull request

 - [x] Provides more friendly options capabilities.
 - [x] Add validation rules for options, and used for Web-UI to prompt user to configure option value;


<!-- Describe the purpose of this pull request. For example: This pull request adds checkstyle plugin.-->

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

ashulin committed:
[api] improve config options

ashulin committed:
[api] option test & new config

ashulin committed:
[api] option check style

ashulin committed:
[api] support complex type option

Hisoka-X commented on seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java:
```diff
@@ -0,0 +1,148 @@
+    public static ReadonlyConfig fromConfig(Config config) {
+        try {
+            return fromMap(JACKSON_MAPPER.readValue(
+                config.root().render(ConfigRenderOptions.concise()),
```
I think use jackson to deserialize config to `ReadonlyConfig` not a cool way. Why just iterate over all values in config, and put into map?

ashulin replied:
Here the `typesafe config` is converted to a json string and parsed into the required map using jackson

Hisoka-X replied:
> Why just iterate over all values in config, and put into map?

My fault, It should be `Why not just iterate over all values in config, and put into map`. If do that, we wouldn't use jackson to serialize and deserialize

ashulin replied:
I have tried to convert `config` without using jackson, but the `config` is more complicated, and there will be many problems in my own implementation. config provides `config.root().render(ConfigRenderOptions.concise())` for converting to json, and it has been well tested.

Hisoka-X commented on seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java:
```diff
@@ -0,0 +1,148 @@
+    public Map<String, String> toMap() {
+        synchronized (this.confData) {
+            Map<String, Object> flatteningMap = flatteningMap(confData);
+            Map<String, String> result = new HashMap<>(flatteningMap.size());
```
`flatteningMap.size()` will convert to array size in map, not map can store data size. If you want set size in hashmap, please use 
```suggestion
            Map<String, String> result = new HashMap<>((float)flatteningMap.size()/0.75 + 1);
```
Or just remove set capacity (recommend).

ashulin replied:
I don't understand, this method doesn't change the size of map keys.

Hisoka-X replied:
Check this: https://cloud.tencent.com/developer/article/1913508

ashulin replied:
You're absolutely right. Thanks

Hisoka-X commented on seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java:
```diff
@@ -0,0 +1,148 @@
+            throw new NullPointerException("Option not be null.");
+        }
+        String[] keys = option.key().split("\\.");
+        synchronized (this.confData) {
```
getOptional use `synchronized` not a great idea. The performance must not good. Maybe you can use cache or something to precompute

ashulin replied:
You have point here.

Hisoka-X commented on seatunnel-api/src/main/java/org/apache/seatunnel/api/configuration/ReadonlyConfig.java:
```diff
@@ -0,0 +1,148 @@
+        } else if (obj instanceof ReadonlyConfig) {
+            Map<String, Object> otherConf = ((ReadonlyConfig) obj).confData;
+
+            for (Map.Entry<String, Object> e : this.confData.entrySet()) {
```
Just use `map.equals` can solve the requirement.

ashulin committed:
[api] deprecated synchronized & simplified equlas

A read-only configuration don't require the 'synchronized' keyword.
hashmap initialcapacity: https://cloud.tencent.com/developer/article/1913508

ashulin committed:
[api] suppress magic number warning

ashulin commented:
@Hisoka-X PTAL

Hisoka-X reviewed (approved):
LGTM
